src/coreclr/src/interop/inc/interoplib.h

```diff
@@ -0,0 +1,108 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _INTEROP_INC_INTEROPLIB_H_
+#define _INTEROP_INC_INTEROPLIB_H_
+
+namespace InteropLibImports
+{
+    // Forward declaration of Runtime calling context.
+    // This class is used by the consuming runtime to pass through details
+    // that may be required during a subsequent callback from the InteropLib.
+    // InteropLib never directly modifies or inspects supplied instances.
+    struct RuntimeCallContext;
+}
+
+namespace InteropLib
+{
+    using OBJECTHANDLE = void*;
+    using OBJECTREF_PROTECTED = void*;
```

Comment:
This makes me worried. This seems to be getting into manually managed code that VM has a lot of infrastructure for to make sure that it is written correctly.

If the interop library is cleanly separated, it needs to be 100% preemptive mode. 

If it needs to be manually managed code where something runs cooperative and something preemptive, etc., it needs to have the VM-style contract to capture and validate it.

Reply:
I agree. I want to chat in person about the issue and hear your thoughts on what options exist. Avoiding this concept in the lib is possible, but does come with a cost. At the moment I was mitigating the risk by putting proper contracts on the call back into the runtime if this type is actually used.

Reply:
Having contracts on the edges helps a bit, but it makes the code hard to review. It is the solution we have used in .NET Framework 1.0. We went through a lot of effort for .NET Framework 2.0 to put the contracts everywhere to make the code auditable.

Reply:
> It is the solution we have used in .NET Framework 1.0. We went through a lot of effort for .NET Framework 2.0 to put the contracts everywhere to make the code auditable.

History wins again. I will take another shot at this.

Reply:
@jkotas Thought on creating a function that the interoplib can call (i.e. assert) if the current thread is Cooperative? This would allow us to assert a contract in the interoplib. There are options, but keeping how to ensure a wrapper is active in the interoplib is my desire.

Reply:
I do not think it is useful to create new style for writing manually managed code. I think that all manually managed code (ie what is running in cooperative mode) should be written in the same style.

Reply:
Thanks for pointing this out @jkotas. Taking the time to rethink it was a good idea.